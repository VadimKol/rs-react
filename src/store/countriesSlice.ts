import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { countries } from '@/common/countries';

import { type RootState } from './store';

export const countriesLice = createSlice({
  name: 'countries',
  initialState: countries,
  reducers: {
    addCountry: (state, action: PayloadAction<string>) => {
      state.push(action.payload);
    },
  },
});

export const { addCountry } = countriesLice.actions;

export const selectCountries = (state: RootState): string[] => state.countries;

export const useCountries = (): string[] => useSelector(selectCountries);
