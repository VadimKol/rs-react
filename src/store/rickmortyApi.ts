import { createApi, fakeBaseQuery } from '@reduxjs/toolkit/query/react';
import { getCharacter, getCharacters } from 'rickmortyapi';

export const rickmortyApi = createApi({
  reducerPath: 'rickmortyApi',
  baseQuery: fakeBaseQuery(),
  endpoints: (builder) => ({
    getCharacters: builder.query({
      queryFn: async ({ page, character }: { page: number; character: { name: string } }) => {
        try {
          const { data, status, statusMessage } = await getCharacters({ page, name: character.name });

          if (status === 200) {
            return { data: { characters: data.results || [], totalPages: data.info?.pages || 0 } };
          }
          if (status === 404) {
            return { data: { characters: [], totalPages: 0 } };
          }

          throw new Error(statusMessage);
        } catch (error) {
          return { error };
        }
      },
    }),
    getCharacter: builder.query({
      queryFn: async (id: number) => {
        try {
          const { data, status, statusMessage } = await getCharacter(id);

          if (status === 200) {
            return { data };
          }
          if (status === 404) {
            return { data: null };
          }

          throw new Error(statusMessage);
        } catch (error) {
          return { error };
        }
      },
    }),
  }),
});

export const { useGetCharactersQuery, useGetCharacterQuery } = rickmortyApi;
