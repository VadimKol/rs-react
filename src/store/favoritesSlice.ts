/* import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';
import { type Character } from 'rickmortyapi';

import { type RootState } from './store';

const initialState: Character[] = [];

export const favoritesSLice = createSlice({
  name: 'favorites',
  initialState,
  reducers: {
    addToFavorites: (state, action: PayloadAction<Character>) => {
      state.push(action.payload);
    },
    removeFromFavorites: (state, action: PayloadAction<Character>) =>
      state.filter((character) => character.id !== action.payload.id),
    clearFavorites: () => [],
  },
});

export const { addToFavorites, removeFromFavorites, clearFavorites } = favoritesSLice.actions;

const selectFavorites = (state: RootState): Character[] => state.favorites;

const isFavorite = (state: RootState, character?: Character): boolean =>
  state.favorites.some((favorite) => favorite.id === character?.id);

export const useFavorites = (): Character[] => useSelector(selectFavorites);

export const useIsFavorite = (character?: Character): boolean =>
  useSelector((state: RootState) => isFavorite(state, character));
 */
