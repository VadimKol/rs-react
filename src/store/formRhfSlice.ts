import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { type FormInfo } from '@/common/types';

import { type RootState } from './store';

const initialState: FormInfo = {
  name: null,
  age: null,
  email: null,
  password: null,
  confirmPassword: null,
  gender: null,
  country: null,
  tc: null,
  image: null,
  strength: '0',
};

export const formRhfSLice = createSlice({
  name: 'formRhf',
  initialState,
  reducers: {
    updateFormRhf: (_, action: PayloadAction<FormInfo>) => action.payload,
  },
});

export const { updateFormRhf } = formRhfSLice.actions;

const selectFormRhfInfo = (state: RootState): FormInfo => state.formRhf;

export const useFormRhfInfo = (): FormInfo => useSelector(selectFormRhfInfo);
