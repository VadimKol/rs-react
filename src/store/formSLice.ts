import { createSlice, type PayloadAction } from '@reduxjs/toolkit';
import { useSelector } from 'react-redux';

import { type FormInfo } from '@/common/types';

import { type RootState } from './store';

export const initialState: FormInfo[] = [];

export const formSLice = createSlice({
  name: 'form',
  initialState,
  reducers: {
    updateForm: (state, action: PayloadAction<FormInfo>) => {
      state.push(action.payload);
    },
  },
});

export const { updateForm } = formSLice.actions;

const selectFormInfo = (state: RootState): FormInfo[] => state.form;

export const useFormInfo = (): FormInfo[] => useSelector(selectFormInfo);
