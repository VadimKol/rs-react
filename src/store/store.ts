import { configureStore } from '@reduxjs/toolkit';

import { formSLice } from './formSLice';

export const store = configureStore({ reducer: { form: formSLice.reducer } });

export type RootState = ReturnType<typeof store.getState>;
