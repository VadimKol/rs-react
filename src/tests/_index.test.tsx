import { render } from '@testing-library/react';
import { Provider } from 'react-redux';

import Page from '@/app//routes/_index';
import { store } from '@/store/store';

import { characters } from './mocks/data';

global.URL.createObjectURL = vi.fn();
global.URL.revokeObjectURL = vi.fn();

const navigate = vi.fn((arg: string): string => arg);

describe('Root route', () => {
  it('renders correctly', () => {
    vi.mock('@remix-run/react', async (importOriginal) => {
      const actual = await importOriginal<object>();
      return {
        ...actual,
        useSearchParams: vi.fn(() => [new URLSearchParams('')]),
        useLoaderData: vi.fn(() => ({ characters, total: 1 })),
        useLocation: vi.fn(() => ({ pathname: '' })),
        useNavigate: vi.fn(() => navigate),
        useNavigation: vi.fn(() => ({ state: 'idle' })),
        Link: vi.fn(({ children }) => <div>{children}</div>),
      };
    });

    const { container } = render(
      <Provider store={store}>
        <Page />
      </Provider>,
    );

    expect(container).toMatchSnapshot();
  });
});
