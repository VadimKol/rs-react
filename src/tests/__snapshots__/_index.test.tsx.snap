// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Root route > renders correctly 1`] = `
<div>
  <main
    class="main"
    role="presentation"
  >
    <section>
      <form
        class="_search_form_5ff5ab"
      >
        <input
          class="_search_5ff5ab"
          id="search"
          placeholder="Search..."
          type="text"
          value=""
        />
        <button
          aria-label="Search-button"
          class="_search_button_5ff5ab"
          type="submit"
        />
      </form>
    </section>
    <section
      class="_results_24fb31"
    >
      <div
        class="_character_24fb31"
      >
        <div
          class="_results_04010f"
        >
          <ul
            class="_characters_04010f"
            role="presentation"
          >
            <li
              class="_characters_item_04010f"
            >
              <div
                class="_card_0086a2"
              >
                <div
                  class="_image_container_de2c00"
                >
                  <div
                    class="_loader_d8e6ec"
                  />
                  <img
                    alt="Character"
                    class="_hidden_de2c00"
                    height="300"
                    src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
                    width="300"
                  />
                </div>
                <div
                  class="_text_container_0086a2"
                >
                  <div>
                    <h2
                      class="_card_title_0086a2"
                    >
                      Rick
                    </h2>
                  </div>
                  <button
                    aria-label="Add to favorites"
                    class="_favorite_99b3bf"
                    name="favorite"
                    type="button"
                    value="false"
                  >
                    ☆
                  </button>
                </div>
              </div>
            </li>
            <li
              class="_characters_item_04010f"
            >
              <div
                class="_card_0086a2"
              >
                <div
                  class="_image_container_de2c00"
                >
                  <div
                    class="_loader_d8e6ec"
                  />
                  <img
                    alt="Character"
                    class="_hidden_de2c00"
                    height="300"
                    src="https://rickandmortyapi.com/api/character/avatar/2.jpeg"
                    width="300"
                  />
                </div>
                <div
                  class="_text_container_0086a2"
                >
                  <div>
                    <h2
                      class="_card_title_0086a2"
                    >
                      Morty
                    </h2>
                  </div>
                  <button
                    aria-label="Add to favorites"
                    class="_favorite_99b3bf"
                    name="favorite"
                    type="button"
                    value="false"
                  >
                    ☆
                  </button>
                </div>
              </div>
            </li>
          </ul>
          <div
            class="_pagination_34bd36"
          >
            <div
              class="_pagination_box_34bd36"
            >
              <div
                class="_pages_34bd36"
              >
                <span>
                  1
                </span>
                -
                <span>
                  1
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </section>
  </main>
</div>
`;
